File: ImageProcessing.java
// ImageProcessing C++ -> Java - Edward Ayala
public class ImageProcessing {

    public static class Pixel {
        private int value;
        private int row;
        private int col;
        public GrayImage img;

        public Pixel() {
            value = 0;
        }

        public Pixel(int value, int row, int col, GrayImage img) {
            this.value = value;
            this.row = row;
            this.col = col;
            this.img = img;
        }

        public int getValue() {
            return value;
        }

        public void setValue(int value) {
            this.value = value;
        }

        public int getRow() {
            return row;
        }

        public void setRow(int row) {
            this.row = row;
        }

        public int getCol() {
            return col;
        }

        public void setCol(int col) {
            this.col = col;
        }

        public void print(){
            System.out.print("\t" + value + "\t");
        }
    }
    public static class GrayImage {
        private int rows;
        private int cols;
        public Pixel [][] pixels;

        public GrayImage( int [][] arr, int rows, int cols) {
            pixels = new Pixel[rows][];
            for (int i = 0; i < rows; i++)
                pixels[i] = new Pixel [cols];
            this.rows = rows;
            this.cols = cols;
            for (int i = 0; i < rows; i++)
            {
                for (int j = 0; j < cols; j++)
                {
                    pixels[i][j] = new Pixel(arr[i][j], i, j, this);
                //  pixels[i][j].setValue(arr[i][j]);
                }
            }
        }

        public Pixel getPixel(int x, int y){
            return pixels[x][y];
        }

        public int countWhitePixels(){
            int n = 0; //The number of white pixels (i.e. 255)

            for (int i = 0; i < rows; i++)
            {
                for (int j = 0; j < cols; j++)
                    if (getPixel(i, j).getValue() == 255)
                        n++;
            }
            return n;
        }

        public void processImage(){
            for (int i = 0; i < rows; i++)
            {
                for (int j = 0; j < cols; j++)
                {
                    if (i + 2 < rows && j + 2 < cols)
                    {
                        int diffPixel = pixels[i][j].getValue() - pixels[i + 2][j + 2].getValue();

                        if (diffPixel < 0)
                        {
                            pixels[i][j].setValue(0);
                        }
                        else
                        {
                            pixels[i][j].setValue(diffPixel);
                        }
                    }
                }
            }
        }

        public int [] getHistogram(){
            int [] histArr = new int [256];
            for (int i = 0; i < rows; i++)
            {
                for (int j = 0; j < cols; j++)
                {
                    int pixVal = pixels[i][j].getValue();
                    for (int count = 0; count < 256; count++)
                    {
                        if (count == pixVal)
                            histArr[count]++;
                    }
                }
            }
            System.out.println("Histogram: ");
            for (int i = 0; i < 256; i++)
                System.out.print(histArr[i] + " ");
            System.out.println();
            return histArr;
        }

        public void print(){
            System.out.println("The pixel values of the image are:");

            for (int i = 0; i < rows; i++)
            {
                for (int j = 0; j < cols; j++)
                {
                    pixels[i][j].print();
                }
                System.out.println();
            }
        }

    public int getCols() {
        return cols;
    }

    public void setCols(int cols) {
        this.cols = cols;
    }

    public int getRows() {
        return rows;
    }

    public void setRows(int rows) {
        this.rows = rows;
    }
    }

    public static void main(String [] args){
        //Create a 2D dynamic array
        int [][] a;
        a = new int [4][];

        for (int i = 0; i < a.length; i++)
            a[i] = new int[5];

        //Input pixel values to a temporary 2D array
        int [][] b = {
            { 221, 184, 178, 84, 135 },
            { 84, 255, 255, 130, 84 },
            { 78, 255, 0, 0, 78 },
            { 84, 130, 255, 130, 84 }
        };

        //Copy to the dynamic array
        for(int i = 0; i < b.length; i++)
        {
            for (int j = 0; j < b[0].length; j++)
            {
                a[i][j] = b[i][j];
            }
        }

        GrayImage gimg = new GrayImage(a, 4, 5);
        gimg.print();
        System.out.println("The number of white pixels is " + gimg.countWhitePixels());
        gimg.getHistogram();
        gimg.processImage();
        gimg.print();
    }
}

File: ImageProcessing.cpp
// Image Processing Assignment - Edward Ayala

#include <iostream>
#include <iomanip>
using namespace std;

class GrayImage;  //Pre-declaration

//Classes
class Pixel
{
public:
	Pixel(int v, int r, int c, GrayImage * m);
	Pixel()
	{
		value = 0;
	}
	int getValue()
	{
		return value;
	}
	void setValue(int val)
	{
		value = val;
	}
	void print();

private:
	int value;
	int rol;
	int col;
	GrayImage * img;
};

//v: pixel value; r: row; c: column; m is a pointer to the image to which this pixel belongs.
Pixel::Pixel(int v, int r, int c, GrayImage * m)
{
	//To be completed
	value = v;
	rol = r;
	col = c;
	img = m;
}

void Pixel::print()
{
	cout << setw(4) << value << " ";
}

// Global Variables
const int BLACK = 0;
const int WHITE = 255;

class GrayImage
{
public:
	GrayImage(int ** arr, int r, int c);
	int countWhitePixels();
	void processImage();
	Pixel * getPixel(int, int);
	int getRows();
	int getCols();
	void print();
	int * getHistogram();

private:
	int rows;
	int cols;

	Pixel *** pixels;  //A 2D array of pointers to pixels (of type Pixel)
};


//Constructor. arr is a 2D array of pixel values. r: rows; c: columns
GrayImage::GrayImage(int ** arr, int r, int c)
{
	//Create a 2D array of pointers to pixels (of type Pixel)
	pixels = new Pixel **[r];

	for (int i = 0; i < r; i++)
		pixels[i] = new Pixel *[c];

	rows = r;
	cols = c;

	for (int i = 0; i < rows; i++)
	{
		for (int j = 0; j < cols; j++)
		{
			pixels[i][j] = new Pixel();
			pixels[i][j]->setValue(arr[i][j]);
		}
	}

}

Pixel * GrayImage::getPixel(int r, int c)
{
	return pixels[r][c];
}

int GrayImage::countWhitePixels()
{
	int n = 0; //The number of white pixels (i.e. 255)

	for (int i = 0; i < rows; i++)
	{
		for (int j = 0; j < cols; j++)
			if (getPixel(i, j)->getValue() == WHITE)
				n++;
	}
	return n;
}
void GrayImage::processImage()
{
	for (int i = 0; i < rows; i++)
	{
		for (int j = 0; j < cols; j++)
		{
			if (i + 2 < rows && j + 2 < cols)
			{
				int diffPixel = pixels[i][j]->getValue() - pixels[i + 2][j + 2]->getValue();

				if (diffPixel < 0)
				{
					pixels[i][j]->setValue(0);
				}
				else
				{
					pixels[i][j]->setValue(diffPixel);
				}
			}
		}
	}
}

int * GrayImage::getHistogram()
{
	int histArr[256] = { 0 };
	for (int i = 0; i < rows; i++)
	{
		for (int j = 0; j < cols; j++)
		{
			int pixVal = pixels[i][j]->getValue();
			for (int count = 0; count < 256; count++)
			{
				if (count == pixVal)
					histArr[count]++;
			}
		}
	}
	cout << "\nHistogram: \n";
	for (int i = 0; i < 256; i++)
	{
		cout << histArr[i] << " ";
	}
	cout << endl;
	return histArr;
}


void GrayImage::print()
{
	cout << "\nThe pixel values of the image are: \n" << endl;

	for (int i = 0; i < rows; i++)
	{
		for (int j = 0; j < cols; j++)
		{
			pixels[i][j]->print();
		}
		cout << endl;
	}
}

int main()
{
	//Create a 2D dynamic array
	int ** a;
	a = new int *[4];

	for (int i = 0; i < 4; i++)
		a[i] = new int[5];

	//Input pixel values to a temporary 2D array
	int b[4][5] = {
		{ 221, 184, 178, 84, 135 },
		{ 84, 255, 255, 130, 84 },
		{ 78, 255, 0, 0, 78 },
		{ 84, 130, 255, 130, 84 }
	};

	//Copy to the dynamic array
	for (int i = 0; i < 4; i++)
	{
		for (int j = 0; j < 5; j++)
		{
			a[i][j] = b[i][j];
		}
	}

	GrayImage gimg(a, 4, 5);
	gimg.print();
	cout << "\nThe number of white pixels is " << gimg.countWhitePixels() << endl;
	gimg.getHistogram();
	gimg.processImage();
	gimg.print();
	return 0;
}


//The pixel values of the image are:
//
// 221  184  178   84  135
//  84  255  255  130   84
//  78  255    0    0   78
//  84  130  255  130   84
//
//The number of white pixels is 4
//
//The pixel values of the image are:
//
// 221  184  100   84  135
//   0  125  171  130   84
//  78  255    0    0   78
//  84  130  255  130   84
//Press any key to continue . . .